tailfs/tailfsimpl/dirfs/dirfs_test.go

```diff
@@ -0,0 +1,353 @@
+		err      error
+	}{
+		{
+			label: "root folder",
+			name:  "",
+			expected: []fs.FileInfo{
+				&shared.StaticFileInfo{
+					Named:      "domain",
+					Sized:      0,
+					ModdedTime: clock.Now(),
+					Dir:        true,
+				},
+			},
+		},
+		{
+			label: "static root folder",
+			name:  "/domain",
+			expected: []fs.FileInfo{
+				&shared.StaticFileInfo{
+					Named:      "remote1",
+					Sized:      0,
+					ModdedTime: clock.Now(),
+					Dir:        true,
+				},
+				&shared.StaticFileInfo{
+					Named:      "remote2",
+					Sized:      0,
+					ModdedTime: clock.Now(),
+					Dir:        true,
+				},
+				&shared.StaticFileInfo{
+					Named:      "remote4",
+					Sized:      0,
+					ModdedTime: clock.Now(),
+					Dir:        true,
+				},
+			},
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			var infos []fs.FileInfo
+			file, err := cfs.OpenFile(ctx, test.name, os.O_RDONLY, 0)
+			if err == nil {
+				defer file.Close()
+				infos, err = file.Readdir(0)
+			}
+			if test.err != nil {
+				if err == nil || !errors.Is(err, test.err) {
+					t.Errorf("expected error: %v   got: %v", test.err, err)
+				}
+			} else {
+				if err != nil {
+					t.Errorf("unable to stat file: %v", err)
+				} else {
+					if len(infos) != len(test.expected) {
+						t.Errorf("wrong number of file infos, want %d, got %d", len(test.expected), len(infos))
+					} else {
+						for i, expected := range test.expected {
+							infosEqual(t, expected, infos[i])
+						}
+					}
+				}
+			}
+		})
+	}
+}
+
+func TestMkdir(t *testing.T) {
+	fs, _, _, _, close := createFileSystem(t)
+	defer close()
+
+	tests := []struct {
+		label string
+		name  string
+		perm  os.FileMode
+		err   error
+	}{
+		{
+			label: "attempt to create root folder",
+			name:  "/",
+		},
+		{
+			label: "attempt to create static root folder",
+			name:  "/domain",
+		},
+		{
+			label: "attempt to create remote",
+			name:  "/domain/remote1",
+		},
+		{
+			label: "attempt to create non-existent remote",
+			name:  "/domain/remote3",
+			err:   os.ErrPermission,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.Mkdir(ctx, test.name, test.perm)
+			if test.err != nil {
+				if err == nil || !errors.Is(err, test.err) {
+					t.Errorf("expected error: %v   got: %v", test.err, err)
+				}
+			} else if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+		})
+	}
+}
+
+func TestRemoveAll(t *testing.T) {
+	fs, _, _, _, close := createFileSystem(t)
+	defer close()
+
+	tests := []struct {
+		label string
+		name  string
+		err   error
+	}{
+		{
+			label: "attempt to remove root folder",
+			name:  "/",
+			err:   os.ErrPermission,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.RemoveAll(ctx, test.name)
+			if err == nil || !errors.Is(err, test.err) {
```

Comment:
We don't need to check nil separately, `!errors.Is(err, test.err)` will cover that.